src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolBlock2.cs

```diff
@@ -0,0 +1,184 @@
+        /// the Slab's ArrayPtr for a slab-block segment. The byte it points to corresponds to Data.Array[0], and in practice you will always
+        /// use the _dataArrayPtr + Start or _dataArrayPtr + End, which point to the start of "active" bytes, or point to just after the "active" bytes.
+        /// </summary>
+        private IntPtr _dataArrayPtr;
+
+        /// <summary>
+        /// The array segment describing the range of memory this block is tracking. The caller which has leased this block may only read and
+        /// modify the memory in this range.
+        /// </summary>
+        public ArraySegment<byte> Data;
+
+        /// <summary>
+        /// This object cannot be instantiated outside of the static Create method
+        /// </summary>
+        protected MemoryPoolBlock2()
+        {
+        }
+
+        /// <summary>
+        /// Back-reference to the memory pool which this block was allocated from. It may only be returned to this pool.
+        /// </summary>
+        public MemoryPool2 Pool { get; private set; }
+
+        /// <summary>
+        /// Back-reference to the slab from which this block was taken, or null if it is one-time-use memory.
+        /// </summary>
+        public MemoryPoolSlab2 Slab { get; private set; }
+
+        /// <summary>
+        /// Convenience accessor
+        /// </summary>
+        public byte[] Array => Data.Array;
+
+        /// <summary>
+        /// The Start represents the offset into Array where the range of "active" bytes begins. At the point when the block is leased
+        /// the Start is guaranteed to be equal to Array.Offset. The value of Start may be assigned anywhere between Data.Offset and
+        /// Data.Offset + Data.Count, and must be equal to or less than End.
+        /// </summary>
+        public int Start { get; set; }
+
+        /// <summary>
+        /// The End represents the offset into Array where the range of "active" bytes ends. At the point when the block is leased
+        /// the End is guaranteed to be equal to Array.Offset. The value of Start may be assigned anywhere between Data.Offset and
+        /// Data.Offset + Data.Count, and must be equal to or less than End.
+        /// </summary>
+        public int End { get; set; }
+
+
+        /// <summary>
+        /// Reference to the next block of data when the overall "active" bytes spans multiple blocks. At the point when the block is
+        /// leased Next is guaranteed to be null. Start, End, and Next are used together in order to create a linked-list of discontiguous 
+        /// working memory. The "active" memory is grown when bytes are copied in, End is increased, and Next is assigned. The "active" 
+        /// memory is shrunk when bytes are consumed, Start is increased, and blocks are returned to the pool.
+        /// </summary>
+        public MemoryPoolBlock2 Next { get; set; }
+
+        ~MemoryPoolBlock2()
```

Comment:
didn't want to implement dispose pattern because finalizer isn't actually disposing, and i didn't want the public api hint that blocks are IDisposable (which they technically aren't)... Also - was concerned that in the finalizer the world is in such an unstable shape - calling log on the finalizer thread from a dying object could end up calling into many object that are already on the road to re-collection